R/modify-records.R

```diff
@@ -0,0 +1,151 @@
+#'   **`remove_records()`** read in the control stream, make any modifications,
+#'   and then save out the updated control stream.
+#'     - `modify_prob_statement()` also returns a character string defining the
+#'     `$PROBLEM` text (see `prob_text` argument).
+#' - **`mod_has_record()`** will return a logical value denoting whether a `bbr`
+#'   model has a given record type.
+NULL
+
+#' @describeIn modify_records Safely read in a `NONMEM` control stream file via
+#' `nmrec`
+#' @keywords internal
+safe_read_ctl <- function(.mod){
+  check_model_object(.mod, c(NM_MOD_CLASS, NMBOOT_MOD_CLASS))
+  mod_path <- get_model_path(.mod)
+  ctl <- tryCatch(nmrec::read_ctl(mod_path), error = identity)
+  if(inherits(ctl, "error")){
+    warning(
+      glue::glue("Could not read control stream file. Reason: {ctl$message}")
+    )
+  }
+  return(ctl)
+}
+
+
+#' @describeIn modify_records Modify or retrieve the `$PROBLEM` statement from
+#' a `NONMEM` control stream file.
+#'
+#' @param prob_text If `NULL` return the current `$PROBLEM` statement. If a
+#'  character string, set the problem statement to that value.
+#' @keywords internal
+modify_prob_statement <- function(.mod, prob_text = NULL){
+  mod_path <- get_model_path(.mod)
+  ctl <- safe_read_ctl(.mod)
+
+  prob_recs <- nmrec::select_records(ctl, "prob")
+  if (length(prob_recs) != 1) {
+    dev_error("select_records should always return a single $PROBLEM record")
+  }
+
+  prob_rec <- prob_recs[[1]]
+  prob_str <- nmrec::get_record_option(prob_rec, "text")
+  if(is.null(prob_str)){
+    # NONMEM permits a bare "$PROB{newline}", in which case there won't be a text option.
+    prob_rec$parse()
+    idx <- purrr::detect_index(
+      prob_rec$values,
+      function(x) inherits(x, "nmrec_option_record_name")
+    )
+    if(idx == 0){
+      dev_error("Unsupported $PROBLEM record format")
+    }
+    if(!is.null(prob_text)){
+      prob_rec$values <- append(prob_rec$values, paste0(" ", prob_text), idx)
+      prob_str_return <- prob_rec$values[[idx + 1]]$format()
+    }else{
+      prob_str_return <- prob_rec$values[[idx]]$format()
+    }
+  }else{
+    if(!is.null(prob_text)){
+      prob_str$value <- prob_text
+    }
+    prob_str_return <- prob_str$value
+  }
+
+  # write control stream and return problem statement
+  nmrec::write_ctl(ctl, mod_path)
+  return(prob_str_return)
+}
+
+
+#' @describeIn modify_records Modify the specified data path in a `NONMEM`
+#' control stream file
+#'
+#' @param data_path Data path to set in a `$DATA` record.
+#'
+#' @keywords internal
+modify_data_path_ctl <- function(.mod, data_path){
+  ctl <- safe_read_ctl(.mod)
+
+  # Get data record
+  data_rec <- nmrec::select_records(ctl, "data")[[1]]
+  data_rec$parse()
+
+  # Overwrite 'filename' option
+  data_rec$values <- purrr::map(data_rec$values, function(data_opt){
+    if(inherits(data_opt, "nmrec_option_pos") && data_opt$name == "filename"){
+      data_opt$value <- data_path
+    }
+    data_opt
+  })
```

Comment:
Rather than the `map`, I think it'd be simpler to use something like (untested)

```R
opt <- get_record_option(..., "filename")
if (is.null(opt)) {
  dev_error(...)
}

opt$value <- data_path
```

Reply:
Addressed in [528a552](https://github.com/metrumresearchgroup/bbr/pull/687/commits/528a5528f838b85a539498757044d8ebb8c48d94)